### DISCUSSION TITLE
Support Japanese character on Node-RED MCU Dashboard

### DISCUSSION THREAD

**Discussion by kitazaki** [phoddie/node-red-mcu#123] [Ideas]
I tried to display Japanese characters on Node-RED MCU Dashboard.

Node-RED flow editor outputs localized flows.json at default, so the name of tags are in Japanese.
English -> Japanese
home -> "ホーム"
default -> "デフォルト"
button -> "ボタン"

It looks like as below.
home -> "???"
default -> "???"
button A -> "???A"
button B -> "???B"

- before
<img src="https://camo.qiitausercontent.com/a8aadbb39de43f0bf9e667d0231d604c78bf9fd5/68747470733a2f2f71696974612d696d6167652d73746f72652e73332e61702d6e6f727468656173742d312e616d617a6f6e6177732e636f6d2f302f37323437392f34306365393635332d666263372d313066612d326265342d3061656132376534396364382e706e67">

I successed to display Japanese characters by creating fonts and changing 
manifest.json
ui_templates.js

- after
<img src="https://camo.qiitausercontent.com/9fd97761358b0be0fcc9a8e15695dea04119b649/68747470733a2f2f71696974612d696d6167652d73746f72652e73332e61702d6e6f727468656173742d312e616d617a6f6e6177732e636f6d2f302f37323437392f32383236623132632d363238392d323934642d616533302d6365366436353237383039662e706e67">

- details on my blog
https://qiita.com/kitazaki/items/a59fad56db2d6ec99ee0

I think that it is useful to change the Piu font from Node-RED flow editor (node-red-mcu-plugin).
Is there any other good idea?!

**Comment by phoddie** [comment]
This looks **very** cool! Nice work. I need a little time to read the (translated) article. Then, let's discuss how to make this convenient for developers.

**Reply by phoddie** [reply]
Apologies for being a little slow. Now that [Moddable SDK 4.5](https://github.com/Moddable-OpenSource/moddable/releases/tag/4.5.0) is out, I should have time to come back to this.

**Reply by phoddie** [reply]
After reading your article, I understand that you are doing the following:

- Generating a font with Shift-JIS characters
- Modifying the Node-RED MCU manifest to include the font
- Modifying the Node-RED MCU dashboard styles to use the font

Using [this recent PR](https://github.com/Moddable-OpenSource/moddable/pull/1317) from @NW-Lab it is possible to add manifests directly to your flow. It eliminates the need to modify the existing manifest to add the font.

The Dashboard running on computers uses CSS for styling. This could be a way to change the font. But, the MCU does not have CSS. We could use Node-RED environment variables instead. For example, "DASHBOARD_FONT" could be an environment variable to select a font. This can be put on the flow, for example. Then we would update the Node_RED MCU dashboard in `ui_template.js` to use the environment variable. Here's how that might look:

```js
	const FONT = RED.util.getSetting(node, 'DASHBOARD_FONT') ?? "Roboto";

	const result = {
		colors: {
			gauge: halfGray,
			group,
			halfGray,
			transparent: TRANSPARENT,
			white: WHITE,
			widget,
			widgetText,
		},
		skins: {},
		styles: {
			chartNoData: new Style({ font:`18px ${FONT}`, color:halfGray, horizontal:"center" }),
			chartX: new Style({ font:`medium 12px ${FONT}`, color:widgetText, horizontal:"center" }),
			chartY: new Style({ font:`medium 12px ${FONT}`, color:widgetText, horizontal:"right" }),
			notification: new Style({ font:`18px ${FONT}`, color:widgetText, horizontal:"left", left:10, right:10, top:10, bottom:10 }),
			keyboard: new Style({ font:`18px ${FONT}`, color:BLACK }),
			textName: new Style({ font:`18px ${FONT}`, color:widgetText, left:5, right:5 }),
			textValue: new Style({ font:`medium 18px ${FONT}`, color:widgetText, left:5, right:5 }),
			textNameLeft: new Style({ font:`18px ${FONT}`, color:widgetText, horizontal:"left", left:10 }),
			textValueLeft: new Style({ font:`medium 18px ${FONT}`, color:widgetText, horizontal:"left", left:10 }),
			textNameRight: new Style({ font:`18px ${FONT}`, color:widgetText, horizontal:"right", right:10 }),
			textValueRight: new Style({ font:`medium 18px ${FONT}`, color:widgetText, horizontal:"right", right:10 }),
			textField: new Style({ font:`medium 18px ${FONT}`, color:[TRANSPARENT,widgetText,widgetText,widget], horizontal:"left", left:10 }),
		},
		textures: {
		}
	};
```

This are just some ideas to make it easier for developers to use different fonts. What do you think?

**Comment by NW-Lab** [comment]
hello

I think it's better to take variables from manifest config.
```
import config from "mc/config";

const FONT = (config.DASHBOARD_FONT) ? config.DASHBOARD_FONT : "Roboto";
```

I'm trying it now, so I'll write another comment.

Thank you.

**Reply by phoddie** [reply]
@NW-Lab – thank you for joining the discussion! We could use configuration values from the manifest here. It is definitely worth exploring.

Some background. The reason I suggested environment variables is because this is a solution for Node-RED. Using a Node-RED tool should be more familiar to Node-RED developers are familiar with environment variables.

**Reply by NW-Lab** [reply]
@phoddie  san

Since I need to specify the FONT in the manifest resource, I thought it would be easier to understand if I changed it in the same place.

Thank you.

**Comment by NW-Lab** [comment]
Hello

I have confirmed that it works, so I pulled request it.
https://github.com/phoddie/node-red-mcu/pull/124

Thank you.

**Comment by kitazaki** [comment]
@phoddie 
I think your idea is great. The update is reasonable for the Node_RED MCU dashboard in ui_template.js to use the environment variable.

@NW-Lab
Your PR works fine for my environment (macOS).

**Reply by phoddie** [reply]
@kitazaki – thank you!

I see the [latest PR](https://github.com/phoddie/node-red-mcu/pull/125) from @NW-Lab implements the support for using the manifest. I think both the manifest and environment approaches are reasonable. We are exploring right now, so I think we could support both.  @NW-Lab, would you be willing to update your PR for that? I think it could be something like:

```js
const FONT = config.DASHBOARD_FONT ?? RED.util.getSetting(node, 'DASHBOARD_FONT')  ?? "Roboto";
```

(Yes, I am a big fan of the [nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing).)

**Reply by phoddie** [reply]
(It turns out that we cannot use `??` here because `RED.util.getStting()` is defined to return the empty string when the environment variable is not present. So, I had to switch it to use `||` instead.

**Comment by kitazaki** [comment]
@phoddie 
For further discussion, I think it is better to prepare some font files for people not familiar to fontbm tool and characters code.

Other bitmap font files are in $MODDABLE/examples/assets/fonts directory and a Japanese scalable font in $MODDABLE/examples/assets/scalablefonts/NotoSans directory.

To avoid pre-create font files, one of option is to add bitmap font files in above directory and another option is to create bitmap font from scalable font in mcconfig build process by defining manifest like below.

```
"resources": {
    "*-mask": [
        {
            "source": "$(MODDABLE)/examples/assets/scalablefonts/NotoSans/NotoSansJP-Regular",
            "size": 18,
            "blocks": [
                "Hiragana",
                "Katakana",
                "Basic Latin"
            ],
            "characters": "亜唖娃阿哀愛挨姶逢葵茜穐悪握渥旭葦芦鯵梓圧斡扱宛姐虻飴絢綾鮎或"
        }
    ]
}
```

**Reply by phoddie** [reply]
@kitazaki – I completely agree that we should make it easy for developers to use different fonts. Pre-encoding them is not ideal. The files are big and we cannot realistically provide every size. Also, it is fun to choose the font for your project -- it gives the UI a unique look.

Using `fontbm` to generate them is great. I have one question for you. In your example above, you use the Unicode "blocks" plus some other characters. In your article, use use the Shift-JIS list of characters. Should we have a similar option in the manifest for Shift-JIS? I know Shift-JIS is not a Unicode block, but maybe we can find some idea to make that work.

**Reply by kitazaki** [reply]
@phoddie 
Unicode block includes CJK characters. (C: Chinese, J: Japanese, K: Korea)
It is possible to define Unicode block if memory available, but it is better to add Shift-JIS definition to "blocks" to save memory use. 
Shift-JIS code set is more useful for Japanese and hopefully want to add EMOJI code set.

**Reply by phoddie** [reply]
Thanks for the information. Strictly speaking Shift-JIS is not a Unicode block. We could have a non-standard Unicode block for that. That may be the easiest to use.

```json
	"resources": {
		"*-mask": [
			{
				"source": "$(MODDABLE)/examples/assets/scalablefonts/OpenSans/OpenSans-Regular",
				"size": 28,
				"blocks": "X-Shift JIS"
			},
```

But, as you note there are other sets developers may want to use. Maybe instead we should provide a way to specify external files with characters in the manifest? We could put common character sets in the Node-RED MCU repository so that developers can reference them easily. This would also allow developers to easily create their own character sets, such as a subset of Shift-JIS or a subset of Emoji (just flags 🇯🇵🇺🇸🇧🇪, for example). It could look like this:

```json
	"resources": {
		"*-mask": [
			{
				"source": "$(MODDABLE)/examples/assets/scalablefonts/OpenSans/OpenSans-Regular",
				"size": 28,
				"charactersFiles": [
					"$(NODEREDMCU)/assets/shift-jis.txt",
					"$(NODEREDMCU)/assets/emjoi.txt"
				]
			},
```

(Note that fonts are black and white (and gray) only.... so Emoji may not be so interesting)

What do you think?

**Reply by kitazaki** [reply]
@phoddie 
I think it is very reasonable !
We prepare characters file to include from manifest.

I want to display colorful EMOJI in future.

**Reply by phoddie** [reply]
> I think it is very reasonable !

Cool, thanks. I'll try add the `"charactersFiles"` support. I think that should be pretty straightforward.

> We prepare characters file to include from manifest.

Perfect. Thank you.

> I want to display colorful EMOJI in future.

Me too. It should actually be possible today using Glyph Designer. But, fontbm doesn't support color. You can see a simple example of color fonts in examples/commodetto/docs in the "text4" case. It uses the file OpenSans-BoldItalic-52.png for the fonts. The file's contents are black. white. and gray but it is a full color PNG image. You can colorize the PNG file and it will display the font in full color. Anyway.. this is a topic for another thread, but actually most of what is needed is present - it is mostly a tools challenge.

**Reply by phoddie** [reply]
The change to `mcconfig` to support `"characterFiles"` is [now live](https://github.com/Moddable-OpenSource/moddable/commit/a6a71b225bd35201ec76b2c354b7595a1be2ed12). Please give it a try.

**Reply by phoddie** [reply]
@kitazaki – did you have a chance to try this out?

**Reply by kitazaki** [reply]
@phoddie 
Sorry for taking a little time to fix another issue in node-red-mcu-plugin environment.
It just works fine for my environment.

```
Starting build process...
Host system check: Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:34 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T8103
MCU Build system check: p1.4.1-beta.1 + #a02cd43 @ m4.5.0-0-gdd55984
HOME directory check: /Users/kitazaki
Creating build environment for platform sim/moddable_two.
Working directory: /Users/kitazaki/tmp/20240320/mcu-plugin-cache/79g0nnznfq
> cd /Users/kitazaki/tmp/20240320/mcu-plugin-cache/79g0nnznfq
Creating build script file...
> /bin/bash ./build.sh
>> mcconfig -d -x localhost:5004 -m -p sim/moddable_two
### Error: $(FONTBM) environment variable not set. Is fontbm installed?
```

added FONTBM environment variable to mcu_plugin.js (just submit pull request for @ralphwetzel )

```
% diff mcu_plugin.js mcu_plugin.js_org
1866d1865
<             "FONTBM": process.env.FONTBM,
```

**Reply by kitazaki** [reply]
@ralphwetzel 

https://github.com/ralphwetzel/node-red-mcu-plugin/pull/48

**Reply by phoddie** [reply]
Great, thanks for confirming.

**Comment by kitazaki** [comment]
closed.

**Reply by phoddie** [reply]
FYI – Now that this is working, I updated the documentation for fontbm support in the manifest to explain `characterFiles` in the "[Selecting Glyphs to Include](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/commodetto/Creating%20fonts%20for%20Moddable%20applications.md#selecting-glyphs-to-include)" section. (The changes should be live later this week)
